### PR TITLE
[candi] various bashible fixes and improvements

### PR DESCRIPTION
**Pull request by RomanenkoDenys**
## Description

1. Added fail with error if docker registry cannot support bearer token auth.
2. Added bashible unit to setup system-wide proxy from deckhouse config (for closed environments).
3. Fixed registry custom ca certificate step.
## Why do we need it, and what problem does it solve?

1. If docker registry cannot suport bearer token auth, bashible fails with curl error, which is not informative.
2. global proxy is needed for containerd in closed environments to pull images from registry.
3. When registry ca certificate changes, we need to restart containerd. 
## Checklist
- [ ] The code is covered by unit tests.
- [x] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: candi
type: chore
summary: |
  Added fail with error if docker registry cannot support bearer token auth.
  Added bashible unit to setup system-wide proxy from deckhouse config (for closed environments).
  Fixed registry custom ca certificate step.
impact_level: low
```

### CODE REVIEW

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @RomanenkoDenys. I've started the workflow for label 'e2e/run/aws'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @RomanenkoDenys. I've started the workflow for label 'e2e/run/gcp'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @RomanenkoDenys. I've started the workflow for label 'e2e/run/gcp'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @RomanenkoDenys. I've started the workflow for label 'e2e/run/aws'...

**Review by yalosev — DISMISSED** [review]

File: candi/bashible/bashbooster/50_deckhouse_registrypackages.sh [hunk 1]
```diff
@@ -41,8 +41,11 @@ bb-rp-is-installed?() {
   fi
 
   AUTH_HEADER="$(curl --retry 3 -sSLi "${SCHEME}://${REGISTRY_ADDRESS}/v2/" | grep -i "www-authenticate")"
-  AUTH_REALM="$(awk -F "," '{split($1,s,"\""); print s[2]}' <<< "${AUTH_HEADER}")"
-  AUTH_SERVICE="$(awk -F "," '{split($2,s,"\""); print s[2]}' <<< "${AUTH_HEADER}" | sed "s/ /+/g")"
+  AUTH_REALM="$(grep -oE 'Bearer realm="http[s]{0,1}://[a-z0-9\.\:\/\-]+"' <<< ${AUTH_HEADER} | cut -d '"' -f2)"
+  AUTH_SERVICE="$(grep -oE 'service="[[:print:]]+"' <<< "${AUTH_HEADER}" | cut -d '"' -f2 | sed 's/ /+/g')"
+  if [ -z ${AUTH_REALM} ]; then
+    bb-exit 1 "cannot parse registry auth header ${AUTH_HEADER}"
```
Comment (apolovov) [inline candi/bashible/bashbooster/50_deckhouse_registrypackages.sh:47]: There should be description, what's wrong with the header. Like 'couldn't find Bearer realm parameter, consider enabling docker realm in your registry'.

File: candi/bashible/bundles/centos/bootstrap.sh.tpl [hunk 2]
```diff
@@ -78,8 +78,12 @@ bb-rp-get-token() {
   fi
 
   AUTH_HEADER="$(curl --retry 3 -k -sSLi "${SCHEME}://${REGISTRY_ADDRESS}/v2/" | grep -i "www-authenticate")"
-  AUTH_REALM="$(awk -F "," '{split($1,s,"\""); print s[2]}' <<< "${AUTH_HEADER}")"
-  AUTH_SERVICE="$(awk -F "," '{split($2,s,"\""); print s[2]}' <<< "${AUTH_HEADER}" | sed "s/ /+/g")"
+  AUTH_REALM="$(grep -oE 'Bearer realm="http[s]{0,1}://[a-z0-9\.\:\/\-]+"' <<< ${AUTH_HEADER} | cut -d '"' -f2)"
+  AUTH_SERVICE="$(grep -oE 'service="[[:print:]]+"' <<< "${AUTH_HEADER}" | cut -d '"' -f2 | sed 's/ /+/g')"
+  if [ -z ${AUTH_REALM} ]; then
+    >&2 echo "cannot parse registry auth header ${AUTH_HEADER}"
```
Comment (apolovov) [inline candi/bashible/bundles/centos/bootstrap.sh.tpl:84]: (duplicate) There should be description, what's wrong with the header. Like 'couldn't find Bearer realm parameter, consider enabling docker realm in your registry'.

File: candi/bashible/bundles/debian/all/000_install_ca_certificates.sh.tpl [hunk 3]
```diff
@@ -23,11 +23,17 @@ fi
 
 {{- if .registry.ca }}
 bb-event-on 'registry-ca-changed' '_update_ca_certificates'
-function _update_ca_certificates() {
+_update_ca_certificates() {
+  bb-flag-set containerd-need-restart
   update-ca-certificates
 }
 
 bb-sync-file /usr/local/share/ca-certificates/registry-ca.crt - registry-ca-changed << "EOF"
```
Comment (apolovov) [inline candi/bashible/bundles/debian/all/000_install_ca_certificates.sh.tpl:31]:
```suggestion
bb-sync-file /usr/local/share/ca-certificates/d8-registry-ca.crt - registry-ca-changed << "EOF"
```
(duplicate) we should have done this before.

File: candi/bashible/common-steps/all/001_set_system_proxy.sh.tpl [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+NO_PROXY={{ .modulesProxy.noProxy | quote }}
+  {{- end }}
+
+bb-event-on 'bb-sync-file-changed' '_restart_containerd'
+
+mkdir -p /etc/systemd/system.conf.d/
+
+bb-sync-file /etc/systemd/system.conf.d/proxy-default-environment.conf - << EOF
+[Manager]
+DefaultEnvironment="HTTP_PROXY=${HTTP_PROXY:-}" "HTTPS_PROXY=${HTTPS_PROXY:-}" "NO_PROXY=${NO_PROXY:-}"
```
Comment (apolovov) [inline candi/bashible/common-steps/all/001_set_system_proxy.sh.tpl:39]:
```suggestion
DefaultEnvironment="HTTP_PROXY=${HTTP_PROXY:-}" "http_proxy=${HTTP_PROXY:-}" "HTTPS_PROXY=${HTTPS_PROXY:-}" "https_proxy=${HTTPS_PROXY:-}" "NO_PROXY=${NO_PROXY:-}" "no_proxy=${NO_PROXY:-}"
```
It turned out, there are weird inconsistency of standarts.
https://about.gitlab.com/blog/2021/01/27/we-need-to-talk-no-proxy/

File: modules/040-node-manager/templates/bashible-apiserver/deployment.yaml [hunk 5]
```diff
@@ -131,6 +132,20 @@ spec:
...
     {{- end }}
+  {{- if hasKey $.Values.global.modules "proxy" }}
+    modulesProxy:
+    {{- if hasKey $.Values.global.modules.proxy "httpProxy" }}
+      httpProxy: {{ $.Values.global.modules.proxy.httpProxy | quote }}
+    {{- end }}
+    {{- if hasKey $.Values.global.modules.proxy "httpsProxy" }}
+      httpsProxy: {{ $.Values.global.modules.proxy.httpsProxy | quote }}
+    {{- end }}
+    {{- $noProxy := list "169.254.169.254" $.Values.global.clusterConfiguration.clusterDomain $.Values.global.clusterConfiguration.podSubnetCIDR $.Values.global.clusterConfiguration.serviceSubnetCIDR }}
```
Comment (apolovov) [inline modules/040-node-manager/templates/bashible-apiserver/deployment.yaml:143]: 👍

File: candi/bashible/bundles/debian/all/000_install_ca_certificates.sh.tpl (hunk 3 above)
Comment (RomanenkoDenys) [reply candi/bashible/bundles/debian/all/000_install_ca_certificates.sh.tpl:31]: This leads to undesirable work on the nodes.

File: candi/bashible/bundles/centos/all/000_install_ca_certificates.sh.tpl [hunk 6]
```diff
@@ -20,11 +20,18 @@ fi
 
 {{- if .registry.ca }}
 bb-event-on 'registry-ca-changed' '_update_ca_certificates'
-function _update_ca_certificates() {
+_update_ca_certificates() {
+  bb-flag-set containerd-need-restart
   update-ca-trust
 }
 
 bb-sync-file /etc/pki/ca-trust/source/anchors/registry-ca.crt - registry-ca-changed << "EOF"
```
[the remark this answers is not in the document]
Comment (RomanenkoDenys) [reply candi/bashible/bundles/centos/all/000_install_ca_certificates.sh.tpl:28]: This leads to undesirable work on the nodes.

File: candi/bashible/bundles/centos/bootstrap.sh.tpl (hunk 2 above)
Comment (RomanenkoDenys) [reply candi/bashible/bundles/centos/bootstrap.sh.tpl]: done

File: candi/bashible/bashbooster/50_deckhouse_registrypackages.sh (hunk 1 above)
Comment (RomanenkoDenys) [reply candi/bashible/bashbooster/50_deckhouse_registrypackages.sh]: done

**Review by apolovov — APPROVED** [review]

**Review by yalosev — APPROVED** [review]

**Pull request merged**